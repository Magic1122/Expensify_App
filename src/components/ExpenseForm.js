import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css';

const now = moment();
console.log(now.format('MMM Do, YYYY'))

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        // console.log(e.target.value)
        const description = e.target.value
        this.setState({ description })
        // console.log(this.state.description)

    }

    onNoteChange = (e) => {
        const note = e.target.value
        this.setState({ note })
        // console.log(this.state.note)
    }

    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState({ amount })
        }
    }

    onDateChange = (createdAt) => {
        createdAt &&
            this.setState({ createdAt })
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            // set error state qual to please provide desc and amount
            this.setState(() => ({ error: 'Please provide description and amount!' }))
        } else {
            this.setState({ error: '' })
            // console.log('Sumbitted')
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }



    render() {
        return (
            <form className='form' onSubmit={this.onSubmit}>
            {this.state.error && <p1 className='form__error'>{this.state.error}</p1>}
                <input type='text'
                    placeholder='Description'
                    autoFocus
                    className='text-input'
                    value={this.state.description}
                    onChange={(e) => this.onDescriptionChange(e)}>
                </input>
                <input type='text'
                    placeholder='Amount'
                    value={this.state.amount}
                    className='text-input'
                    onChange={this.onAmountChange}></input>
                <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null
                    onDateChange={this.onDateChange} // PropTypes.func.isRequired
                    focused={this.state.calendarFocused} // PropTypes.bool
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    placeholder='Add a note for your expense (optional)'
                    className='textarea'
                    value={this.state.note}
                    onChange={(e) => this.onNoteChange(e)}></textarea>
                <div>
                <button className='button' type='submit'>Save Expense</button>
                </div>
            </form>
        )
    }
}

export default ExpenseForm;