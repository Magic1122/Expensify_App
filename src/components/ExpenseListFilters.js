import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'


class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null,

    }

    onDatesChange = ({ startDate, endDate }) => {
        console.log(startDate, endDate)
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }

    onFocusChange = (calendarFocused) => {
        this.setState({ calendarFocused })
    }

    render() {
        return (<div className='content-container'>
            <div className='input-group'>
                <div className='input-group__item'>
                <input className='text-input' type='text' placeholder='Search expenses' value={this.props.filter.text} onChange={(e) => this.props.dispatch(setTextFilter(e.target.value))} /></div>
                <div className='input-group__item'>
                    <select className='select' value={this.props.filter.sortBy} onChange={(e) => e.target.value == 'date' ? this.props.dispatch(sortByDate()) : this.props.dispatch(sortByAmount())}>
                        <option value='date'>Date</option>
                        <option value='amount'>Amount</option>
                </select></div>
                <div className='input-group__item'><DateRangePicker
                    startDate={this.props.filter.startDate}
                    endDate={this.props.filter.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                /></div>
            </div>


        </div>)
    }
}


const mapStateToProps = state => {
    return {
        filter: state.filter
    }
}


export default connect(mapStateToProps)(ExpenseListFilters);
