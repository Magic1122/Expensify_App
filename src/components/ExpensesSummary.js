import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';


export const ExpensesSummary = (props) => {

    const expensesCount = props.expenses.length
    const expensesTotal = selectExpensesTotal(props.expenses)
    const expenseMessage = expensesCount <= 1 ? 'expense' : 'expenses';

    return (
    <div>
        <h1>Viewing {expensesCount} {expenseMessage} totalling {numeral(expensesTotal / 100).format('$0,0.00')}</h1>
    </div>
);
    }

const mapStateToProps = state => {
    return {
        expenses: selectExpenses(state.expenses, state.filter)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);