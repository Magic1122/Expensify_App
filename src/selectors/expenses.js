import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    console.log('STARTDATE', startDate, 'ENDDATE', endDate)
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        console.log(startDateMatch)
        console.log(endDateMatch)
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy == 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}

export default getVisibleExpenses;
