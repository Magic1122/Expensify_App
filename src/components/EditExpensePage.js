import React from 'react';
import { connect } from 'react-redux';
import ExpenseFrom from './ExpenseForm';
import { startRemoveExpense, startEditExpense, editExpense, removeExpense } from '../actions/expenses';


const EditExpensePage = (props) => {
    console.log(props)
    return (
    <div>
        <ExpenseFrom
        expense={props.expense}
        onSubmit={(expense) => {
            props.dispatch(startEditExpense(props.expense.id, expense))
            props.history.push('/')
        }} />
        <button type='button' onClick={() => props.dispatch(startRemoveExpense({ id: props.expense.id }))
        && props.history.push('/')}>Remove</button>
    </div>
)
    }

const mapStateToProps = (state, props) => {
    console.log('STATE', state, 'PROPS', props)
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }

};

export default connect(mapStateToProps)(EditExpensePage);
