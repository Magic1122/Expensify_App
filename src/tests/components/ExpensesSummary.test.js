import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render an expense total item', () => {

    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} /* id={expenses[1].id} description={expenses[1].description} amount={expenses[1].amount} createdAt={expenses[1].createdAt} */ />)

    expect(wrapper).toMatchSnapshot();
})

test('should render multiple total item', () => {

    const wrapper = shallow(<ExpensesSummary expenses={expenses} /* id={expenses[1].id} description={expenses[1].description} amount={expenses[1].amount} createdAt={expenses[1].createdAt} */ />)

    expect(wrapper).toMatchSnapshot();
})

