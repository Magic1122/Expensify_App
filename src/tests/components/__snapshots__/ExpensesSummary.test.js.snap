// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expense total item 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $1.95
  </h1>
</div>
`;

exports[`should render multiple total item 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $1,141.95
  </h1>
</div>
`;
